pkg/oauth2/token_source.go

```diff
@@ -65,26 +68,46 @@ func (c *passwordCredentialsTokenSource) token(now func() time.Time) (*oauth2.To
 		if tok.RefreshToken == c.refreshToken.RefreshToken {
 			return tok, nil
 		}
-	} else {
-		tok, err = c.cfg.PasswordCredentialsToken(c.ctx, c.username, c.password)
+
+		err = c.setRefreshToken(tok)
 		if err != nil {
-			return nil, fmt.Errorf("password credentials token source failed: %v", err)
+			return nil, err
 		}
 
-		c.accessTokenSource = c.cfg.TokenSource(c.ctx, tok)
+		return tok, nil
 	}
 
+	return c.passwordCredentialsToken()
+}
+
+func (c *passwordCredentialsTokenSource) passwordCredentialsToken() (*oauth2.Token, error) {
```

Comment:
This shouldn’t happen too often, but it would be good to know when it does. Let’s add a counter metric to understand how often this happens.

Reply:
agreed :+1: added a metric, PTAL